File: src/solid/open-closed.ts
// one enterprise pattern explanations as introduction
enum Color {
  red = 'red',
  green = 'green',
  blue = 'blue',
}

enum Size {
  small = 'small',
  medium = 'medium',
  large = 'large',
}

class Product {
  public name: string;
  public color: Color;
  public size: Size;
  constructor(name: string, color: Color, size: Size) {
    this.name = name;
    this.color = color;
    this.size = size;
  }
}

// specification
interface Specification {
  isSatisfied: (p: Product) => boolean;
}
class ColorSpecification implements Specification {
  public color: Color;
  constructor(color: Color) {
    this.color = color;
  }

  isSatisfied(item: Product) {
    return item.color === this.color;
  }
}

class SizeSpecification implements Specification {
  public size: Size;
  constructor(size: Size) {
    this.size = size;
  }

  isSatisfied(item: Product) {
    return item.size === this.size;
  }
}

// if we want to combine specifications for filter we can add a combinator
class AndSpecification {
  public specs: Specification[];
  constructor(...specs: Specification[]) {
    this.specs = specs;
  }

  isSatisfied(item: Product) {
    return this.specs.every((x) => x.isSatisfied(item));
  }
}

let apple = new Product('Apple', Color.green, Size.small);
let tree = new Product('Tree', Color.green, Size.large);
let house = new Product('House', Color.blue, Size.large);

let products = [apple, tree, house];

class BetterFilter {
  filter(items: Product[], spec: Specification) {
    return items.filter((x) => spec.isSatisfied(x));
  }
}

let bf = new BetterFilter();
console.log(`Green products (new):`);
for (let p of bf.filter(products, new ColorSpecification(Color.green))) {
  console.log(` * ${p.name} is green`);
}

console.log(`Large and green products:`);
let spec = new AndSpecification(
  new SizeSpecification(Size.large),
  new ColorSpecification(Color.green)
);
for (let p of bf.filter(products, spec)) {
  console.log(` * ${p.name} is large and green`);
}

File: python/solid/open-closed.py
from __future__ import annotations
from abc import ABC, abstractmethod
from typing import Any
from enum import Enum

class Color(Enum):
    red = 'red'
    green = 'green'
    blue = 'blue'

class Size(Enum):
    small = 'small'
    medium = 'medium'
    large = 'large'

class Product:
    def __init__(self, name: str, color: Color, size: Size):
        self.name = name
        self.color = color
        self.size = size

class Specification:
    @abstractmethod
    def is_satisfied(self, item: Any) -> bool:
        pass
    
    def __and__(self, other):
        return AndSpecification(self, other)

class SizeSpecification(Specification):
    def __init__(self, size: Size):
        self.size = size
    
    def is_satisfied(self, item: Product) -> bool:
        return item.size == self.size
    
class ColorSpecification(Specification):
    def __init__(self, color: Color):
        self.color = color
    
    def is_satisfied(self, item: Product) -> bool:
        return item.color == self.color

class AndSpecification(Specification):
    def __init__(self, *specs: list[Specification]):
        self.specs = specs
    
    def is_satisfied(self, item: Product) -> bool:
        return all(map(lambda spec: spec.is_satisfied(item), self.specs))

class Filter(ABC):
    @abstractmethod
    def filter(self, items: list[Product], spec: Specification) -> list[Product]:
        pass
    
class SimpleFilter(Filter):
    def filter(self, items: list[Product], spec: Specification) -> list[Product]:
        return list(filter(lambda item: spec.is_satisfied(item), items))

apple = Product('Apple', Color.green, Size.small)
tree = Product('Tree', Color.green, Size.large)
house = Product('House', Color.blue, Size.large)

products = [apple, tree, house]

simple_filter = SimpleFilter()
print('Green products (new):')
green = ColorSpecification(Color.green)
for p in simple_filter.filter(products, green):
    print(f' - {p.name} is green')

print('Large and green products:')
spec = AndSpecification(
    ColorSpecification(Color.green), 
    SizeSpecification(Size.large)
  )

for p in simple_filter.filter(products, spec):
    print(f' - {p.name} is large and green')